File: student_Banasthali_Nivedita/trees/problem7.cpp
/*
Kth smallest element in BST
*/

#include <iostream>
#include <stack>

using namespace std;

struct TreeNode {
    int val;
    TreeNode* left = NULL;
    TreeNode* right = NULL;
    TreeNode(int x) : val(x) {}
};

int kthSmallest(TreeNode* root, int k) {
    stack<TreeNode*> st;
    TreeNode* curr = root;
    
    while (curr != NULL || !st.empty()) {
        while (curr != NULL) {
            st.push(curr);
            curr = curr->left;
        }
        
        curr = st.top();
        st.pop();
        
        k--;
        if (k == 0) {
            return curr->val;
        }
        
        curr = curr->right;
    }
    
    return -1; // In case k is invalid or tree is empty
}

int main() {
    // Create the binary search tree
    TreeNode* root = new TreeNode(5);
    root->left = new TreeNode(3);
    root->right = new TreeNode(6);
    root->left->left = new TreeNode(2);
    root->left->right = new TreeNode(4);
    root->left->left->left = new TreeNode(1);

    int k = 3;
    int result = kthSmallest(root, k);
    cout << "The " << k << "th smallest element is: " << result << endl;

    return 0;
}

File: ShrutiVarshney_BanasthaliVidyapith_TreeAssignment/TreeAssignmentQuestions/src/treeassignmentquestions/Problem7.java
package treeassignmentquestions;
public class Problem7 
{
    private static int count;
    private static int result;

    public static int kthSmallest(TreeNode root, int k) 
    {
        count = 0;
        result = 0;
        inorderTraversal(root, k);
        return result;
    }

    private static void inorderTraversal(TreeNode node, int k) 
    {
        if (node == null) 
        {
            return;
        }

        inorderTraversal(node.left, k);

        count++;
        if (count == k) 
        {
            result = node.val;
            return;
        }

        inorderTraversal(node.right, k);
    }

    public static void main(String[] args)
    {
        // Create example tree 1: [3, 1, 4, null, 2]
        TreeNode root1 = new TreeNode(3);
        root1.left = new TreeNode(1);
        root1.right = new TreeNode(4);
        root1.left.right = new TreeNode(2);

        // Find the kth smallest element
        int k1 = 1;
        int result1 = kthSmallest(root1, k1);

        // Print the result
        System.out.println("The " + k1 + "th smallest element is: " + result1);

        // Create example tree 2: [5, 3, 6, 2, 4, null, null, 1]
        TreeNode root2 = new TreeNode(5);
        root2.left = new TreeNode(3);
        root2.right = new TreeNode(6);
        root2.left.left = new TreeNode(2);
        root2.left.right = new TreeNode(4);
        root2.left.left.left = new TreeNode(1);

        // Find the kth smallest element
        int k2 = 3;
        int result2 = kthSmallest(root2, k2);

        // Print the result
        System.out.println("The " + k2 + "th smallest element is: " + result2);
    }
}
